addon/model-data.js

```diff
@@ -1,183 +1,212 @@
-
-    if (value !== oldValue) {
-      // Add the new value to the changed attributes hash; it will get deleted by
-      // the 'didSetProperty' handler if it is no different from the original value
-      this._attributes[key] = value;
-
-      if (key in this._inFlightAttributes) {
-        originalValue = this._inFlightAttributes[key];
-      } else {
-        originalValue = this._data[key];
-      }
-      // If we went back to our original value, we shouldn't keep the attribute around anymore
-      if (value === originalValue) {
-        delete this._attributes[key];
-      }
+  getOrCreateNestedModelData(key, modelName, id, internalModel) {
+    let nestedModelData = this._nestedModelDatas[key];
+    if (!nestedModelData) {
+      nestedModelData = this._nestedModelDatas[
+        key
+      ] = this.createNestedModelData(modelName, id, internalModel);
     }
+    return nestedModelData;
   }
 
-  getAttr(key) {
-    // TODO IGOR DAVID investigate why attributes would be null
-    if (this._attributes && key in this._attributes) {
-      return this._attributes[key];
-    } else if (this._inFlightAttributes && key in this._inFlightAttributes) {
-      return this._inFlightAttributes[key];
-    } else {
-      return this._data[key];
+  createNestedModelData(modelName, id, internalModel) {
+    return new M3ModelData(
+      modelName,
+      id,
+      null,
+      this.storeWrapper,
+      this.store,
+      internalModel
+    );
+  }
+
+  destroyNestedModelData(key) {
+    let nestedModelData = this._nestedModelDatas[key];
+    if (nestedModelData) {
+      // destroy
+      delete this._nestedModelDatas[key];
     }
   }
 
+  hasNestedModelData(key) {
+    return !!this._nestedModelDatas[key];
+  }
+
+  setAttr(key, value) {
+    this._data[key] = value;
+  }
+
+  getAttr(key) {
+    return this._data[key];
+  }
+
   hasAttr(key) {
-    return (
-      key in this._attributes ||
-      key in this._inFlightAttributes ||
-      key in this._data
-    );
+    return key in this._data;
+  }
+
+  getResourceIdentifier() {
+    let { modelName, clientId, id } = this.internalModel;
+
+    return {
+      id,
+      clientId,
+      type: modelName,
+    };
   }
 
-  get _attributes() {
-    if (this.__attributes === null) {
-      this.__attributes = Object.create(null);
+  /**
+   *
+   * @param updates
+   * @param nestedCallback a callback for updating the data of a nested model-data instance
+   * @returns {Array}
+   * @private
+   */
+  _mergeUpdates(updates, nestedCallback, calculateChanges = true) {
+    let data = this._data;
+
+    let changedKeys;
+    if (calculateChanges) {
+      changedKeys = [];
     }
-    return this.__attributes;
+
+    if (!updates) {
+      // no changes
+      return changedKeys;
+    }
+
+    let updatedKeys = Object.keys(updates);
+
+    for (let i = 0; i < updatedKeys.length; i++) {
+      let key = updatedKeys[i];
+      let newValue = updates[key];
+
+      if (isEqual(data[key], newValue)) {
```

Comment:
I'm unsure that we need `isEqual` here vs `===`: https://github.com/emberjs/ember.js/blob/db34024bae2caa06c4ae845d1dc3261a02e35b69/packages/ember-runtime/lib/is-equal.js#L43-L53